src/aws-manager.js

```diff
@@ -326,65 +326,77 @@ class AwsManager {
             this.provisionerId,
             'http://taskcluster.net/fake-provisioner-base-url',
             this.pubKey,
-            worker.workerType
+            worker.workerType,
+            this
         );
       } else {
-        launchSpecs = worker.testLaunchSpecs();
+        launchSpecs = worker.testLaunchSpecs(this);
       }
     } catch (err) {
       returnValue.canLaunch = false;
-      returnValue.reasons.push(err);
+      if (err.code === 'InvalidLaunchSpecifications') {
+        returnValue.reasons = err.reasons;
+      } else {
+        returnValue.reasons.push(err.toString());
+      }
       return returnValue;
     }
 
-    await Promise.all(worker.regions.map(async r => {
-      if (!this.ec2[r.region]) {
+    await Promise.all(Object.keys(launchSpecs).map(async r => {
```

Comment:
I'd rather iterate over the data we expect instead of the data that we got.  Maybe we should add an assertion that the launchSpec at the expected region/zone/type actually is an object